zipkin-collector/rabbitmq/src/main/java/zipkin/collector/rabbitmq/RabbitMQCollector.java

```diff
@@ -0,0 +1,202 @@
+
+    @Override public Builder storage(StorageComponent storage) {
+      this.delegate.storage(storage);
+      return this;
+    }
+
+    @Override public Builder sampler(CollectorSampler sampler) {
+      this.delegate.sampler(sampler);
+      return this;
+    }
+
+    @Override public Builder metrics(CollectorMetrics metrics) {
+      this.metrics = checkNotNull(metrics, "metrics").forTransport("rabbitmq");
+      this.delegate.metrics(this.metrics);
+      return this;
+    }
+
+    public Builder addresses(List<String> addresses) {
+      this.addresses = convertAddresses(addresses);
+      return this;
+    }
+
+    public Builder concurrency(int concurrency) {
+      this.concurrency = concurrency;
+      return this;
+    }
+
+    public Builder connectionFactory(ConnectionFactory connectionFactory) {
+      this.connectionFactory = checkNotNull(connectionFactory, "connectionFactory");
+      return this;
+    }
+
+    /**
+     * Queue zipkin spans will be consumed from. Defaults to "zipkin-spans".
+     */
+    public Builder queue(String queue) {
+      this.queue = checkNotNull(queue, "queue");
+      return this;
+    }
+
+    @Override public RabbitMQCollector build() {
+      return new RabbitMQCollector(this);
+    }
+  }
+
+  final String queue;
+  final LazyInit connection;
+
+  RabbitMQCollector(Builder builder) {
+    this.queue = builder.queue;
+    this.connection = new LazyInit(builder);
+  }
+
+  @Override
+  public RabbitMQCollector start() {
+    connection.get();
+    return this;
+  }
+
+  @Override
+  public CheckResult check() {
+    try {
+      CheckResult failure = connection.failure.get();
+      if (failure != null) return failure;
+      return CheckResult.OK;
+    } catch (RuntimeException e) {
+      return CheckResult.failed(e);
+    }
+  }
+
+  @Override public void close() throws IOException {
+    connection.close();
+  }
+
+  /** Lazy creates a connection and a queue before starting consumers */
+  static final class LazyInit extends LazyCloseable<Connection> {
+    final Builder builder;
+    final AtomicReference<CheckResult> failure = new AtomicReference<>();
+
+    LazyInit(Builder builder) {
+      this.builder = builder;
+    }
+
+    @Override
+    protected Connection compute() {
+      Connection connection;
+      try {
+        connection = builder.connectionFactory.newConnection(builder.addresses);
+        connection.createChannel().queueDeclare(builder.queue, true, false, false, null);
+      } catch (IOException | TimeoutException e) {
+        throw new IllegalStateException("Unable to establish connection to RabbitMQ server", e);
+      }
+      Collector collector = builder.delegate.build();
+      CollectorMetrics metrics = builder.metrics;
+
+      for (int i = 0; i < builder.concurrency; i++) {
+        String name = RabbitMQSpanConsumer.class.getName() + i;
+        try {
+          // TODO: Is channel 1-1 with consumer? if not re-use the same channel for each consumer
```

Comment:
Generally speaking, yes. Channels should not be shared among threads. From the [Channel JavaDoc](https://www.rabbitmq.com/releases/rabbitmq-java-client/current-javadoc/com/rabbitmq/client/Channel.html):

> Channel instances must not be shared between threads. Applications should prefer using a Channel per thread instead of sharing the same Channel across multiple threads. While some operations on channels are safe to invoke concurrently, some are not and will result in incorrect frame interleaving on the wire. Sharing channels between threads will also interfere with Publisher Confirms. As such, applications need to use a Channel per thread.

Reply:
thx for the ref